crypto-keys/src/test/java/com/palantir/crypto2/keys/ChainedAsyncKeyStorageStrategyTest.java

```diff
@@ -0,0 +1,113 @@
+
+    private static final Executor EXECUTOR = MoreExecutors.directExecutor();
+    private static final CompletableFuture<Void> VOID = new CompletableFuture<>();
+    private static final String KEY = "key";
+
+    private KeyMaterial keyMaterial;
+    private AsyncKeyStorageStrategy failingStrategy;
+    private AsyncKeyStorageStrategy successfulStrategy;
+    private ChainedAsyncKeyStorageStrategy chained;
+
+    @Before
+    public void before() {
+        keyMaterial = mock(KeyMaterial.class);
+        failingStrategy = mock(AsyncKeyStorageStrategy.class);
+        successfulStrategy = mock(AsyncKeyStorageStrategy.class);
+
+        when(successfulStrategy.put(KEY, keyMaterial)).thenReturn(VOID);
+        when(failingStrategy.put(KEY, keyMaterial)).thenReturn(VOID);
+
+        when(successfulStrategy.remove(KEY)).thenReturn(VOID);
+        when(failingStrategy.remove(KEY)).thenReturn(VOID);
+
+        when(successfulStrategy.get(KEY)).thenReturn(CompletableFuture.completedFuture(keyMaterial));
+        when(failingStrategy.get(KEY)).thenThrow(IllegalStateException.class);
+
+        chained = new ChainedAsyncKeyStorageStrategy(EXECUTOR, successfulStrategy, failingStrategy);
+    }
+
+    @Test
+    public void testAllPutsCalled() {
+        chained.put(KEY, keyMaterial).join();
+
+        verify(successfulStrategy).put(KEY, keyMaterial);
+        verify(failingStrategy).put(KEY, keyMaterial);
+        verifyNoMoreInteractions(successfulStrategy, failingStrategy);
+    }
+
+    @Test
+    public void testGetSucceeds() {
+        assertThat(chained.get(KEY).join(), is(keyMaterial));
+
+        verify(successfulStrategy).get(KEY);
+        verifyNoMoreInteractions(successfulStrategy, failingStrategy);
+    }
+
+    @Test
+    public void testFailedGetIgnored() {
+        chained = new ChainedAsyncKeyStorageStrategy(EXECUTOR, failingStrategy, successfulStrategy);
+
+        assertThat(chained.get(KEY).join(), is(keyMaterial));
+
+        verify(failingStrategy).get(KEY);
+        verify(successfulStrategy).get(KEY);
+        verifyNoMoreInteractions(successfulStrategy, failingStrategy);
+    }
+
+    @Test
+    public void testAllStrategiesFail() {
+        chained = new ChainedAsyncKeyStorageStrategy(EXECUTOR, failingStrategy);
+
+        assertThatExceptionOfType(CompletionException.class)
+                .isThrownBy(() -> chained.get(KEY).join())
+                .withCauseInstanceOf(RuntimeException.class)
+                .withMessageContaining(String.format(
+                        "Unable to get key material using any of the provided strategies: %s",
+                        ImmutableList.of(failingStrategy.getClass().getCanonicalName())));
+    }
+
+    @Test
+    public void testAllDeletesCalled() {
+        chained.remove(KEY).join();
+
+        verify(successfulStrategy).remove(KEY);
+        verify(failingStrategy).remove(KEY);
+        verifyNoMoreInteractions(successfulStrategy, failingStrategy);
+    }
+
```

Comment:
worth testing/asserting what happens when you cancel the future? e.g., do outstanding put/delete requests still get sent or not?